cadc-data-ops-fits/src/main/java/org/opencadc/fits/slice/WCSCutoutUtil.java

```diff
@@ -0,0 +1,128 @@
+ *  names of its contributors may        de ses  participants ne peuvent
+ *  be used to endorse or promote        être utilisés pour approuver ou
+ *  products derived from this           promouvoir les produits dérivés
+ *  software without specific prior      de ce logiciel sans autorisation
+ *  written permission.                  préalable et particulière
+ *                                       par écrit.
+ *
+ *  This file is part of the             Ce fichier fait partie du projet
+ *  OpenCADC project.                    OpenCADC.
+ *
+ *  OpenCADC is free software:           OpenCADC est un logiciel libre ;
+ *  you can redistribute it and/or       vous pouvez le redistribuer ou le
+ *  modify it under the terms of         modifier suivant les termes de
+ *  the GNU Affero General Public        la “GNU Affero General Public
+ *  License as published by the          License” telle que publiée
+ *  Free Software Foundation,            par la Free Software Foundation
+ *  either version 3 of the              : soit la version 3 de cette
+ *  License, or (at your option)         licence, soit (à votre gré)
+ *  any later version.                   toute version ultérieure.
+ *
+ *  OpenCADC is distributed in the       OpenCADC est distribué
+ *  hope that it will be useful,         dans l’espoir qu’il vous
+ *  but WITHOUT ANY WARRANTY;            sera utile, mais SANS AUCUNE
+ *  without even the implied             GARANTIE : sans même la garantie
+ *  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+ *  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+ *  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+ *  General Public License for           Générale Publique GNU Affero
+ *  more details.                        pour plus de détails.
+ *
+ *  You should have received             Vous devriez avoir reçu une
+ *  a copy of the GNU Affero             copie de la Licence Générale
+ *  General Public License along         Publique GNU Affero avec
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.fits.slice;
+
+import ca.nrc.cadc.dali.Circle;
+import ca.nrc.cadc.dali.Interval;
+import ca.nrc.cadc.dali.Polygon;
+import ca.nrc.cadc.dali.Range;
+import ca.nrc.cadc.dali.Shape;
+import ca.nrc.cadc.wcs.exceptions.NoSuchKeywordException;
+
+import java.util.List;
+
+import nom.tam.fits.Header;
+import nom.tam.fits.HeaderCardException;
+import org.opencadc.soda.server.Cutout;
+
+
+public class WCSCutoutUtil {
+    public static long[] getBounds(final Header header, final Cutout cutout)
+            throws HeaderCardException, NoSuchKeywordException {
```

Comment:
This only does cutout of one type: the first one found in hard coded order. In SODA, the caller can specify cutout along multiple axes and they should be combined, eg
```
CIRCLE=12 34 0.1 & BAND=200e-9 400e-9 & TIME=54321.5 54321.6 & POL=I & POL=Q
```
(extra space for clarity) in a 5-D cube. You can treat the axes as orthogonal so what I saw of the spatial cutouts didn't obviously make me think this was what you had in mind... the caom2 cutout code does each axis separately and recombines the long[] pixel ranges into the single n-D pixel cutout action so the FITSCutout and subclass(es) approach can work, but not as simply as this.

Reply:
I had to dig very deep to see how this is intended to work - I had missed the change in FITSCutout.getBounds behaviour where it returns a long[NAXIS] rather than a long[n] or long[0]. The javadocs don't consistently describe the contract of that method. 

Specifically in PolygonCutout, lines ~238-248: there is an assumption in the array copy and fill that the spatial axes are 1 and 2 but that doesn't have to be the case. Energy could be 1 and spatial could be 2 and 3. So that needs to be generalised and WCSCutoutUtilTest needs tests with axes in different orders.

EnergyCutout lines~161-175 do not make that assumption and are a better model of how to put the clipped bounds into the right place in the entireBounds array.

Reply:
It's a bit of a shame that the clipped bounds -> entire bounds happens in multiple Cutout<T> implementations; as such, each should have a test where the axis in question is in an atypical place in the order...

PolygonCutoutTest probably needs to be updated with a test where spatial axes are not 1,2... and if you don't have it, I'd also have a test where DEC axis comes before RA (we do have some data like that in the archive, hence the CoordSys.swappedAxes) because you might need to take that into account when doing clipped -> entire bounds.

The EnergyCutoutTest(s) all seem to have energyAxis == 3 (which is typical for cube). I would at least add a test for 1-D spectra so that energyAxis == 1 (also typical).